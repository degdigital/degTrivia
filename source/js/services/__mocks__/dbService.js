const dbService = jest.genMockFromModule('./dbService');

function submitAnswer() {
	return Promise.resolve(true);
}

function getLeaderboardData() {
	return Promise.resolve({
		game: [
			{name: 'Anna', score: '1500000'},
			{name: 'Aaron', score: '140'},
			{name: 'Ryan', score: '42'}
		],
		event: [
			{name: 'Aaron', score: '1500001'},
			{name: 'Aaron', score: '200'},
			{name: 'Ryan', score: '43'}
		]
	})
}

dbService.submitAnswer = submitAnswer;
dbService.getLeaderboardData = getLeaderboardData;

export default dbService;