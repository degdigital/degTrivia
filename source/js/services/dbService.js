import firebase from '@firebase/app';
import '@firebase/database';

const dbService = function() {

	let db = null;
	const nodeNames = {
		events: 'events',
		series: 'series',
		games: 'games',
		pendingPlayers: 'pendingPlayers'
	};
	const defaultNodesToGet = [
		'events',
		'series',
		'games'
	];

	function init() {
		db = firebase.database();
	}

	function getDb() {
		return db;
	}

	function getInitialData(nodes = defaultNodesToGet) {
		return new Promise(async(resolve, reject) => {
			let output = {};
			const data = await Promise.all(nodes.map(node => db.ref(node).once('value').then(snapshot => snapshot.val())));
			nodes.map((node, index) => output[node] = data[index]);
			resolve(output);
		});
	}

	function getCurrentEventId() {
		return db.ref('currentEvent').once('value').then(snapshot => snapshot.val());
	}

	function getEvent(eventAlias) {
		return db.ref('events').orderByChild('alias').equalTo(eventAlias).once('value').then(snapshot => snapshot.val());
	}

	function createPendingPlayer(playerVals, eventKey, userId) {
		const pendingPlayerKey = db.ref(nodeNames.pendingPlayers).push().key;
		
		return db.ref(`/${nodeNames.pendingPlayers}/${pendingPlayerKey}`).update(formattedPlayerVals)
			.then(() => pendingPlayerKey);
	}

	async function getNextGameTime() {
		const currentEventId = await getCurrentEventId();
		const games = await db.ref('games').orderByChild('event').equalTo(currentEventId).once('value').then(snapshot => snapshot.val());
		if (!games) {
			return Promise.resolve(null);
		}
		const nextGameTime = new Date(Object.keys(games).map(gameId => games[gameId].startTime).sort()[0]);
		return Promise.resolve(nextGameTime);
	}

	function submitAnswer(questionId, choiceId, playerId) {
		if (questionId && choiceId && playerId){
			return db.ref(`answers/${questionId}/responses/${choiceId}`).update({
				[playerId]: true
			});
		}
	}

	function getActiveGameData(gameId) {
		return new Promise((resolve, reject) => {
			db.ref(`games/${gameId}`).once('value', snapshot => {
				if (snapshot.exists()) {
					resolve(snapshot.val());	
				} else {
					reject();
				}
			});
		});
	}

	function getLeaderboardData() {
		return db.ref('leaderboardCurrent').once('value').then(snapshot => {
			if (snapshot.val()) {
				return snapshot.val();
			} else {
				return {};
			}
		})
	}

	return {
		init,
		getDb,
		getInitialData,
		getCurrentEventId,
		getEvent,
		getNextGameTime,
		submitAnswer,
		getLeaderboardData,
		getActiveGameData
	};

};

const instance = dbService();

export default instance;