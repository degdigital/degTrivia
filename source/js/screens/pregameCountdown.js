import {replaceContent} from '../utils/domUtils.js';
import countdown from '../components/countdown.js';
import dbService from '../services/dbService.js';
import { format as formatDate, isSameDay, differenceInMilliseconds } from 'date-fns';

const countdownThreshold = 900000; //15 minutes in milliseconds
const countdownDataAttr = 'data-countdown';

const pregameCountdown = function(element) {
    let countdownInst;

    function renderNoNextGameMessage() {
        return `<p>No other games are currently scheduled.</p>`;
    }

    function renderCountdown() {
        return `<div class="countdown next-game__time" ${countdownDataAttr}></div>`;
    }

    function renderNextGameTime(nextGameTime) {
        console.log(nextGameTime);
        const timeText = formatDate(nextGameTime, 'h:mma'); 

        return `<time datetime="${nextGameTime.toISOString()}" class="next-game__time">
                    ${timeText}
                </time>`;
    }

    function renderNextGameTimeIntro(showCountdown) {
        let introHtml = 'Next Game';
        if(showCountdown) {
            introHtml += '<br />Begins In';
        }
        return introHtml;
    }

    function renderNextGame(nextGameTime, showCountdown) {
        return `
            <div class="next-game">
                <div class="next-game__primary">
                    <img src="/images/clock.svg" alt="Clock" class="next-game__clock-image" />
                    <h1 class="next-game__title">
                        <div class="next-game__title-intro">${renderNextGameTimeIntro(showCountdown)}</div>
                        ${showCountdown ? renderCountdown() : renderNextGameTime(nextGameTime)}
                    </h1>
                    <div class="event-hashtag next-game__event-hashtag">#CNXTRIVIA</div>
                </div>
                <p class="next-game__message">Speed counts! Answer as quickly as you can.</p>
            </div>
        `;
    }

    function startCountdown(countdownEl, nextGameTime) {
        const timeTilNextGame = differenceInMilliseconds(nextGameTime, new Date());
       
        countdownInst = countdown({
            containerElement: countdownEl, 
            format: 'mm:ss',
            onComplete: onCountdownComplete
        });
        countdownInst.start(timeTilNextGame, 'milliseconds');
    }

    function onCountdownComplete() {
        //Not sure if we still need to do anything here
    }

    function isGameTimeWithinCountdownThreshold(date) {
        const difference = differenceInMilliseconds(date, new Date());
        return (difference <= countdownThreshold && difference >= 0); 
    }

    async function renderContent() {
        const nextGameTime = await dbService.getNextGameTime();
        if (nextGameTime) {
            const showCountdown = isGameTimeWithinCountdownThreshold(nextGameTime);

            return renderNextGame(nextGameTime, showCountdown);

        } 
        
        return renderNoNextGameMessage();
    }

	async function render() {
        const nextGameTime = await dbService.getNextGameTime();
        const showCountdown = nextGameTime ? 
            isGameTimeWithinCountdownThreshold(nextGameTime) :
            false;

        const contentHtml = nextGameTime ?
            renderNextGame(nextGameTime, showCountdown) :
            renderNoNextGameMessage()

        const html = `
            <div class="pregame">
                ${contentHtml}
            </div>
        `;

        replaceContent(element, html);

        if(showCountdown) {
            const countdownEl = element.querySelector(`[${countdownDataAttr}]`);
            startCountdown(countdownEl, nextGameTime);
        }
	}

	return {
        render,
        teardown: () => {
            if (countdownInst) {
                countdownInst.stop();
            }
        }
	};

};

export default pregameCountdown;