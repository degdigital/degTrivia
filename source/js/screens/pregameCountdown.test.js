import pregameCountdown from './pregameCountdown';
import dbService from '../services/dbService';

jest.mock('../services/dbService', () => {
    return {
        getNextGameTime: jest
            .fn()
            .mockResolvedValue(null)
            .mockResolvedValueOnce(new Date('May 5, 2020 12:00:00'))
    }
});

describe('pregameCountdown', () => {
    let element;

    beforeEach(() => {
        document.body.innerHTML = '';
        element = document.createElement('div');
	    document.body.appendChild(element);
    });

    it('renders', async () => {
        await pregameCountdown({element}).render();
        expect(element).toBeTruthy();

        const countdownContainer = element.querySelector('.countdown-container');
        expect(countdownContainer).toBeTruthy();

        const nextTimeEl = element.querySelector('.next-game-time');
        expect(nextTimeEl).toBeTruthy();
    })

    it('renders nothing if no next game', async () => {
        await pregameCountdown({element}).render();
        expect(document.body.innerHTML).toEqual(`<div></div>`
        );
    });

});