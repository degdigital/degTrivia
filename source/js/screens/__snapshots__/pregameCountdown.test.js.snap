// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pregameCountdown renders countdown if next game starts exactly 15 minutes from now: countdown 1`] = `
<div>
  
            
  <div
    class="pregame"
  >
    
                
            
    <div
      class="next-game"
    >
      
                
      <div
        class="next-game__primary"
      >
        
                    
        <img
          alt="Clock"
          class="next-game__clock-image"
          src="/images/clock.svg"
        />
        
                    
        <h1
          class="next-game__title"
        >
          
                        
          <div
            class="next-game__title-intro"
          >
            Next Game
            <br />
            Begins In
          </div>
          
                        
          <div
            class="countdown next-game__time"
            data-countdown=""
          />
          
                    
        </h1>
        
                    
        <div
          class="event-hashtag next-game__event-hashtag"
        >
          #CNXTRIVIA
        </div>
        
                
      </div>
      
                
      <p
        class="next-game__message countdown-rotating-copy"
      />
      
            
    </div>
    
        
            
  </div>
  
        
</div>
`;

exports[`pregameCountdown renders countdown if next game starts less than 15 minutes from now: countdown 1`] = `
<div>
  
            
  <div
    class="pregame"
  >
    
                
            
    <div
      class="next-game"
    >
      
                
      <div
        class="next-game__primary"
      >
        
                    
        <img
          alt="Clock"
          class="next-game__clock-image"
          src="/images/clock.svg"
        />
        
                    
        <h1
          class="next-game__title"
        >
          
                        
          <div
            class="next-game__title-intro"
          >
            Next Game
            <br />
            Begins In
          </div>
          
                        
          <div
            class="countdown next-game__time"
            data-countdown=""
          />
          
                    
        </h1>
        
                    
        <div
          class="event-hashtag next-game__event-hashtag"
        >
          #CNXTRIVIA
        </div>
        
                
      </div>
      
                
      <p
        class="next-game__message countdown-rotating-copy"
      />
      
            
    </div>
    
        
            
  </div>
  
        
</div>
`;

exports[`pregameCountdown renders game ready message after countdown completes: game ready message 1`] = `
<div>
  
            
  <div
    class="pregame"
  >
    
                
            
    <div
      class="next-game"
    >
      
                
      <div
        class="next-game__primary"
      >
        
                    
        <img
          alt="Clock"
          class="next-game__clock-image"
          src="/images/clock.svg"
        />
        
                    
        <h1
          class="next-game__title"
        >
          
                        
          <div
            class="next-game__title-intro"
          >
            Next Game
            <br />
            Begins In
          </div>
          
                        
          <div
            class="countdown next-game__time"
            data-countdown=""
          />
          
                    
        </h1>
        
                    
        <div
          class="event-hashtag next-game__event-hashtag"
        >
          #CNXTRIVIA
        </div>
        
                
      </div>
      
                
      <p
        class="next-game__message countdown-rotating-copy"
      />
      
            
    </div>
    
        
            
  </div>
  
        
</div>
`;

exports[`pregameCountdown renders game ready message if next game start time has passed: game ready message 1`] = `
<div>
  
            
  <div
    class="pregame"
  >
    
                
            
    <div
      class="next-game"
    >
      
                
      <div
        class="next-game__primary"
      >
        
                    
        <img
          alt="Clock"
          class="next-game__clock-image"
          src="/images/clock.svg"
        />
        
                    
        <h1
          class="next-game__title"
        >
          
                        
          <div
            class="next-game__title-intro"
          >
            Next Game
          </div>
          
                        
          <time
            class="next-game__time"
            datetime="1980-09-22T04:59:00.000Z"
          >
            
                11:59pm
            
          </time>
          
                    
        </h1>
        
                    
        <div
          class="event-hashtag next-game__event-hashtag"
        >
          #CNXTRIVIA
        </div>
        
                
      </div>
      
                
      <p
        class="next-game__message countdown-rotating-copy"
      />
      
            
    </div>
    
        
            
  </div>
  
        
</div>
`;

exports[`pregameCountdown renders game time message if next game starts more than 15 minutes from now: game time message 1`] = `
<div>
  
            
  <div
    class="pregame"
  >
    
                
            
    <div
      class="next-game"
    >
      
                
      <div
        class="next-game__primary"
      >
        
                    
        <img
          alt="Clock"
          class="next-game__clock-image"
          src="/images/clock.svg"
        />
        
                    
        <h1
          class="next-game__title"
        >
          
                        
          <div
            class="next-game__title-intro"
          >
            Next Game
          </div>
          
                        
          <time
            class="next-game__time"
            datetime="1980-09-22T05:20:00.000Z"
          >
            
                12:20am
            
          </time>
          
                    
        </h1>
        
                    
        <div
          class="event-hashtag next-game__event-hashtag"
        >
          #CNXTRIVIA
        </div>
        
                
      </div>
      
                
      <p
        class="next-game__message countdown-rotating-copy"
      />
      
            
    </div>
    
        
            
  </div>
  
        
</div>
`;

exports[`pregameCountdown renders no next game message if there is not a next game: no next game message 1`] = `
<div>
  
            
  <div
    class="pregame"
  >
    
                
    <p>
      No other games are currently scheduled.
    </p>
    
            
  </div>
  
        
</div>
`;
